### PR TITLE
[clang-cl][AST] Fix auto NTTP MSVC 1920+ mangling for pointer types

### PR DESCRIPTION
**Pull request by MaxEW707**
https://godbolt.org/z/G1K8Wszn9 for reference.

Starting with MSVC 1920+, VS2019, C++17 auto NTTP now adds `M <type>` to the mangled name to avoid name collisions with different deduced types.
This PR fixes pointers. Pointers to members will be fixed in an upcoming PR.

Here is a small example. The godbolt has more thorough examples.
```
template<auto>
struct AutoParmTemplate
{
    AutoParmTemplate() {}
};

int i;

int main()
{
    // MSVC 1916: ??0?$AutoParmTemplate@$1?i@@3HA@@QEAA@XZ
    // MSVC 1929: ??0?$AutoParmTemplate@$MPEAH1?i@@3HA@@QEAA@XZ
    // Clang: ??0?$AutoParmTemplate@$1?i@@3HA@@QEAA@XZ
    AutoParmTemplate<&i> x;
}
```

### CODE REVIEW

**Comment by MaxEW707** [comment]
Friendly ping @rnk @efriedma-quic @zmodem

**Review by efriedma-quic — APPROVED** [review]
LGTM

Your commit message mentions variables, function pointers, and member function pointers...  is there anything else we need to handle?  Null pointers?  Integers?

**Review by zmodem — COMMENTED** [review]
Sorry for the delay (please feel free to ping earlier). This looks good to me.

Will llvm/lib/Demangle/MicrosoftDemangle.cpp also need to be updated? (That could be done separately)

**Comment by MaxEW707** [comment]
> Will llvm/lib/Demangle/MicrosoftDemangle.cpp also need to be updated? (That could be done separately)

Possibly. I didn't realize the llvm-project implemented demangling as well. I'll take a look at the demangle code after I merge this PR and get a separate PR up if it needs updating.

> Your commit message mentions variables, function pointers, and member function pointers... is there anything else we need to handle? Null pointers? Integers?

auto return functions that have NTTP parameters need fixing: https://github.com/llvm/llvm-project/issues/92204

Integers were fixed in this commit here: https://github.com/llvm/llvm-project/commit/83286a1a8f059d1664b64341854676a36a85cecd

I need to check null pointers.

I haven't checked C++20 float point or literal class types either.

**Pull request merged**
